## Import JSON for NHMS - Mammals

In [16]:
import pandas as pd

# Load the JSON file into a DataFrame
# df = pd.read_json('normal_test.json')
# df = pd.read_json('mammals_10k_rand_1.json')
# Print the head of the DataFrame
# print(df.head())

!python -m json.tool mammals_10k_rand_1.json

Extra data: line 1 column 400417 (char 400416)


## Explore physical descriptions
For sort by: most relevant in mammals collection only

In [4]:
print(df['physicalDescriptions'].head(20))

0                         [Skeleton, Skin, Skull, Male]
1                                      [Fluid, Unknown]
2                         [Skeleton, Skin, Skull, Male]
3                                [Skin, Skull, Unknown]
4                                      [Fluid, Unknown]
5           [Antler or Horn, pair cut from skull, Male]
6                                      [Fluid, Unknown]
7     [7 g, 99 mm, 48 mm, 11 mm, 12 mm, 42 mm, Skin,...
8                  [Partial Skull, Ramus only, Unknown]
9     [57.7 g, 236 mm, 117 mm, 33 mm, 12 mm, Skeleto...
10                        [Skeleton, Skin, Skull, Male]
11                                     [Fluid, Unknown]
12                        [Skeleton, Skin, Skull, Male]
13                               [Skin, Skull, Unknown]
14                                     [Fluid, Unknown]
15          [Antler or Horn, pair cut from skull, Male]
16                                     [Fluid, Unknown]
17    [7 g, 99 mm, 48 mm, 11 mm, 12 mm, 42 mm, S

In [5]:
# create dataset where physicalDescriptions strings includes at least one number. Print head and total count. 
df['physicalDescriptions'] = df['physicalDescriptions'].astype(str)
df = df[df['physicalDescriptions'].str.contains(r'\d', na=False)]
print(df['physicalDescriptions'].head(20))
print(df['physicalDescriptions'].count())

7     ['7 g', '99 mm', '48 mm', '11 mm', '12 mm', '4...
9     ['57.7 g', '236 mm', '117 mm', '33 mm', '12 mm...
17    ['7 g', '99 mm', '48 mm', '11 mm', '12 mm', '4...
19    ['57.7 g', '236 mm', '117 mm', '33 mm', '12 mm...
20                                 ['210 cm', 'Female']
22    ['50 g', '150 mm', '51 mm', '20 mm', '31 mm', ...
24    ['13 g', '125 mm', '99 mm', '16 mm', '9 mm', '...
25    ['1098 mm', '497 mm', '166 mm', 'Skin', 'Skull...
29         ['61 mm', '55 mm', '10 mm', 'Fluid', 'Male']
30    ['47 g', '104 mm', '10 mm', '17 mm', '18 mm', ...
32    ['380 mm', '380 mm', '150 mm', 'Skin', 'Skull'...
36    ['58 g', '270 mm', '158 mm', '25 mm', '17 g', ...
37    ['39 g', '100 mm', '8 mm', '13 mm', '17 mm', '...
38    ['27 g', '75 mm', '0 mm', '14 mm', '18 mm', '5...
39    ['23 g', '144 mm', '45 mm', '18 mm', '15 mm', ...
40    ['110 g', '373 mm', '203 mm', '34 mm', '22 mm'...
42    ['400 g', '470 mm', '216 mm', '57 mm', '65 mm'...
43    ['92 mm', '0 mm', '17 mm', '23 mm', '68 mm

# With 405 records out of 1010 with numbers in physicalDescriptions
we can see that 40% of the records have numbers in the physicalDescriptions field.

Repeat this process for the first number that is followed by ' mm' and add it to a new column called "length", and turn that number into an integer.
If ' mm' is not found but a ' cm' is found, convert the ' cm' to ' mm' and add it to the "length" column.
astly, if there are no instances of ' g' or ' mm' in the physicalDescription, add a NaN to the weight and length columns.

In [6]:
# for each row and in physicalDescription, identify the first number that is followed by ' g' and add it to a new column called "weight", and turn that number into an integer.
df['weight'] = df['physicalDescriptions'].str.extract(r'(\d+) g', expand=False)
df['weight'] = df['weight'].fillna(0).astype(int)
print(df.tail(20))

                                     id  \
966   ld1-1643413365956-1643413435944-0   
970   ld1-1643413365956-1643413439576-1   
972   ld1-1643414459528-1643414465800-0   
973   ld1-1643414459528-1643414459763-0   
976   ld1-1643414459528-1643414485792-0   
977   ld1-1643414459528-1643414464250-0   
978   ld1-1643413935791-1643413957238-1   
979   ld1-1643414459528-1643414459647-0   
980   ld1-1643414459528-1643414471636-1   
982   ld1-1643414459528-1643414463488-1   
984   ld1-1643413365956-1643413440219-0   
988   ld1-1643413365956-1643413438198-1   
991   ld1-1643414459528-1643414459977-1   
994   ld1-1643414459528-1643414465373-1   
1000  ld1-1643413365956-1643413440287-1   
1001  ld1-1643414459528-1643414463936-1   
1003  ld1-1643413365956-1643413440165-1   
1006  ld1-1643414459528-1643414463135-1   
1018  ld1-1643414459528-1643414464206-0   
1019  ld1-1643414459528-1643414465420-0   

                                      title  \
966   Physeter macrocephalus Linnaeus, 1758   
97

In [7]:
# for each row in physicalDescription, identify the first number that is followed by ' mm' and add it to a new column called "length", and turn that number into an integer.
df['length'] = df['physicalDescriptions'].str.extract(r'(\d+) mm', expand=False)
df['length'] = df['length'].astype(float)
print(df.tail(20))

                                     id  \
966   ld1-1643413365956-1643413435944-0   
970   ld1-1643413365956-1643413439576-1   
972   ld1-1643414459528-1643414465800-0   
973   ld1-1643414459528-1643414459763-0   
976   ld1-1643414459528-1643414485792-0   
977   ld1-1643414459528-1643414464250-0   
978   ld1-1643413935791-1643413957238-1   
979   ld1-1643414459528-1643414459647-0   
980   ld1-1643414459528-1643414471636-1   
982   ld1-1643414459528-1643414463488-1   
984   ld1-1643413365956-1643413440219-0   
988   ld1-1643413365956-1643413438198-1   
991   ld1-1643414459528-1643414459977-1   
994   ld1-1643414459528-1643414465373-1   
1000  ld1-1643413365956-1643413440287-1   
1001  ld1-1643414459528-1643414463936-1   
1003  ld1-1643413365956-1643413440165-1   
1006  ld1-1643414459528-1643414463135-1   
1018  ld1-1643414459528-1643414464206-0   
1019  ld1-1643414459528-1643414465420-0   

                                      title  \
966   Physeter macrocephalus Linnaeus, 1758   
97

In [8]:
# for each row in physicalDescription, identify the first number that is followed by ' mm' and add it to a new column called "length", and turn that number into an integer.
df['length'] = df['physicalDescriptions'].str.extract(r'(\d+) mm', expand=False)
df['length'] = df['length'].astype(float)
print(df.tail(20))

                                     id  \
966   ld1-1643413365956-1643413435944-0   
970   ld1-1643413365956-1643413439576-1   
972   ld1-1643414459528-1643414465800-0   
973   ld1-1643414459528-1643414459763-0   
976   ld1-1643414459528-1643414485792-0   
977   ld1-1643414459528-1643414464250-0   
978   ld1-1643413935791-1643413957238-1   
979   ld1-1643414459528-1643414459647-0   
980   ld1-1643414459528-1643414471636-1   
982   ld1-1643414459528-1643414463488-1   
984   ld1-1643413365956-1643413440219-0   
988   ld1-1643413365956-1643413438198-1   
991   ld1-1643414459528-1643414459977-1   
994   ld1-1643414459528-1643414465373-1   
1000  ld1-1643413365956-1643413440287-1   
1001  ld1-1643414459528-1643414463936-1   
1003  ld1-1643413365956-1643413440165-1   
1006  ld1-1643414459528-1643414463135-1   
1018  ld1-1643414459528-1643414464206-0   
1019  ld1-1643414459528-1643414465420-0   

                                      title  \
966   Physeter macrocephalus Linnaeus, 1758   
97

In [9]:
# for physicalDescription rows that have missing values for length, if the first number found in the physicalDescription is followed by ' cm', multiply that number by 10 and add it to the length column.
# and if the first number found in the physicalDescription is followed by ' m', multiply that number by 1000 and add it to the length column.
df['length'] = df['length'].fillna(0)
df['length'] = df['length'] + df['physicalDescriptions'].str.extract(r'(\d+) cm', expand=False).fillna(0).astype(float) * 10
df['length'] = df['length'] + df['physicalDescriptions'].str.extract(r'(\d+) m', expand=False).fillna(0).astype(float) * 1000


In [10]:
# for rows that have weight = 0, look in physicalDescription for numbers that are followed by ' kg' and multiply that number by 1000 and add it to the weight column.
df['weight'] = df['weight'] + df['physicalDescriptions'].str.extract(r'(\d+) kg', expand=False).fillna(0).astype(float) * 1000


In [11]:
import numpy as np

df['weight'].replace(0, np.nan, inplace=True)
df['length'].replace(0, np.nan, inplace=True)

print(df[['weight', 'length']].tail(20))

      weight    length
966      NaN   11000.0
970      NaN  345345.0
972      6.0  296296.0
973      4.0       NaN
976   5000.0    1590.0
977      NaN  107107.0
978      NaN       NaN
979     24.0  115115.0
980      2.0  294294.0
982      NaN   70070.0
984     38.0       NaN
988     37.0  116116.0
991     10.0   83083.0
994      4.0  403403.0
1000    19.0  187187.0
1001     4.0   94094.0
1003     1.0   76076.0
1006   170.0  500500.0
1018     NaN   71071.0
1019    21.0  199199.0


/var/folders/pc/j_zmpc9s0g9frv2jzfv1l51c0000gn/T/ipykernel_4866/1142042403.py:3: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['weight'].replace(0, np.nan, inplace=True)
/var/folders/pc/j_zmpc9s0g9frv2jzfv1l51c0000gn/T/ipykernel_4866/1142042403.py:4: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always be

# Attempt 1 of confidence interval for rest of population   

In [12]:
from scipy import stats;
import numpy as np

# Sample size
n = len(df)

# Population size
N = 500000

# Sample mean and standard deviation
sample_mean = df['weight'].mean()
sample_std = df['weight'].std(ddof=1)

# Standard error of the mean
sem = sample_std / np.sqrt(n)

# Confidence level

confidence_level = 0.95

# Degrees of freedom
df_degrees = n - 1

# t-score for the confidence level
t_score = stats.t.ppf((1 + confidence_level) / 2, df_degrees)

# Margin of error
margin_of_error = t_score * sem

# Confidence interval
confidence_interval = (sample_mean - margin_of_error, sample_mean + margin_of_error)

print(f"Confidence Interval: {confidence_interval}")

Confidence Interval: (np.float64(292.9837245874031), np.float64(1929.5586482939525))


This means that based on your sample, you can be 95% confident that the true average weight of the entire population lies between 292.98 and 1929.55 (units are whatever weight is measured in, like grams or kilograms).


In [13]:
# Sample mean and standard deviation for length
sample_mean_length = df['length'].mean()
sample_std_length = df['length'].std(ddof=1)

# Standard error of the mean for length
sem_length = sample_std_length / np.sqrt(n)

# t-score for the confidence level (already defined as t_score)
# Margin of error for length
margin_of_error_length = t_score * sem_length

# Confidence interval for length
confidence_interval_length = (sample_mean_length - margin_of_error_length, sample_mean_length + margin_of_error_length)

print(f"Confidence Interval for Length: {confidence_interval_length}")

Confidence Interval for Length: (np.float64(153904.4484164148), np.float64(196257.22191325552))


“I’m 95% sure the average length of everything in the population is between 153,904.45 and 196,257.22.”